massa-pos-exports/src/types.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2022 MASSA LABS <info@massa.net>
+
+use std::collections::{BTreeMap, VecDeque};
+
+use bitvec::prelude::*;
+use massa_models::{
+    prehash::{Map, Set},
+    Address, Amount, Slot,
+};
+
+pub struct PoSFinalState {
+    /// contiguous cycle history. Front = newest.
+    pub cycle_history: VecDeque<CycleInfo>,
+
+    /// latest final slot
+    pub last_final_slot: Slot,
+
+    /// coins to be credited at the end of the slot
+    pub deferred_credits: BTreeMap<Slot, Map<Address, Amount>>,
+}
+
+pub struct CycleInfo {
+    /// cycle number
+    pub cycle: u64,
+
+    /// whether the cycle is complete (all slots final)
+    pub complete: bool,
+
+    /// number of rolls each staking address has
+    pub roll_counts: Map<Address, u64>,
+
+    /// random seed bits of all slots in the cycle so far
+    pub rng_seed: BitVec<Lsb0, u8>,
+
+    /// Per-address production statistics
+    pub production_stats: Map<Address, ProductionStats>,
```

Comment:
Maybe merge `roll_counts` and `production_stats` into one `per_address_info: Map<Address, AddressCycleInfo>` ?

Reply:
I agree with greg on that one, after starting the `PoSChanges` implementation in active history I've met that need. Will be using that on my PR for now but we should discuss this.

Reply:
What I'm currently using:
```rust
#[derive(Default, Debug, Clone)]
pub struct PoSAddressInfo {
    /// set deferred credits indexed by target slot (can be set to 0 to cancel some, in case of slash)
    /// ordered structure to ensure slot iteration order is deterministic
    pub deferred_credits: BTreeMap<Slot, Amount>,

    /// updated production statistics
    pub production_stats: ProductionStats,

    /// new rolls counts
    pub roll_changes: u64,
}

#[derive(Default, Debug, Clone)]
pub struct PoSChanges {
    /// extra block seed bits added
    pub seed_bits: BitVec<Lsb0, u8>,

    /// addresses info
    pub addresses_info: Map<Address, PoSAddressInfo>,
}
```

Reply:
FI: I rolled back to the current PoSChanges definition, but using `Map<Address, ProductionStats>` instead of `ProductionStats`. It was a mistake in the RFC, it has been updated.